<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>cardio.preprocessing_scripts package &mdash; Cardio 01 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="cardio.tests package" href="cardio.tests.html" />
    <link rel="prev" title="cardio package" href="cardio.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Cardio
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">cardio</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="cardio.html">cardio package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="cardio.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">cardio.preprocessing_scripts package</a></li>
<li class="toctree-l4"><a class="reference internal" href="cardio.tests.html">cardio.tests package</a></li>
<li class="toctree-l4"><a class="reference internal" href="cardio.vis.html">cardio.vis package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="cardio.html#module-cardio">Module contents</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="readme_link.html">README Metabolander</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme_link.html#example-event-parameter"></a></li>
<li class="toctree-l1"><a class="reference internal" href="use_cases.html">Use Cases</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_stories.html">User Stories</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Cardio</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">cardio</a></li>
          <li class="breadcrumb-item"><a href="cardio.html">cardio package</a></li>
      <li class="breadcrumb-item active">cardio.preprocessing_scripts package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/cardio.preprocessing_scripts.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cardio-preprocessing-scripts-package">
<h1>cardio.preprocessing_scripts package<a class="headerlink" href="#cardio-preprocessing-scripts-package" title="Permalink to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-cardio.preprocessing_scripts.load_data">
<span id="cardio-preprocessing-scripts-load-data-module"></span><h2>cardio.preprocessing_scripts.load_data module<a class="headerlink" href="#module-cardio.preprocessing_scripts.load_data" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.load_data.load_data">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.load_data.</span></span><span class="sig-name descname"><span class="pre">load_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sheets_url</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/load_data.html#load_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.load_data.load_data" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-cardio.preprocessing_scripts.norm_filt">
<span id="cardio-preprocessing-scripts-norm-filt-module"></span><h2>cardio.preprocessing_scripts.norm_filt module<a class="headerlink" href="#module-cardio.preprocessing_scripts.norm_filt" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.norm_filt.combine_metamicro">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.norm_filt.</span></span><span class="sig-name descname"><span class="pre">combine_metamicro</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">metadata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metabs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">micros</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/norm_filt.html#combine_metamicro"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.norm_filt.combine_metamicro" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.norm_filt.count_to_abundance">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.norm_filt.</span></span><span class="sig-name descname"><span class="pre">count_to_abundance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/norm_filt.html#count_to_abundance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.norm_filt.count_to_abundance" title="Permalink to this definition"></a></dt>
<dd><p>Transform microbiome sequencing count data to relative abundance with
Center-log ratio (CLR) transformation.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.norm_filt.filter_sparse">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.norm_filt.</span></span><span class="sig-name descname"><span class="pre">filter_sparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percent</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/norm_filt.html#filter_sparse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.norm_filt.filter_sparse" title="Permalink to this definition"></a></dt>
<dd><p>filters sparse features (microbes or metabolites) so that we only consider
those present in at least (percent) % of samples</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.norm_filt.load_data">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.norm_filt.</span></span><span class="sig-name descname"><span class="pre">load_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sheets_url</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/norm_filt.html#load_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.norm_filt.load_data" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-cardio.preprocessing_scripts.trimdata">
<span id="cardio-preprocessing-scripts-trimdata-module"></span><h2>cardio.preprocessing_scripts.trimdata module<a class="headerlink" href="#module-cardio.preprocessing_scripts.trimdata" title="Permalink to this heading"></a></h2>
<p>This script reads csv files, calculates microbe relative abundance and alpha
diversity, filters out rare features (present in &gt; some % samples).</p>
<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.trimdata.basic_filtering">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.trimdata.</span></span><span class="sig-name descname"><span class="pre">basic_filtering</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">microbiome_columns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metabolome_columns</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/trimdata.html#basic_filtering"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.trimdata.basic_filtering" title="Permalink to this definition"></a></dt>
<dd><p>Drop patients missing most measurements, replace zeros with NaN, get
separate dfs for microbiome/metabolome</p>
<blockquote>
<div><dl class="simple">
<dt>Parameters:</dt><dd><p>df(Pandas dataframe): entire dataset
microbiome_columns(column index object):
metabolome_columns(column index object):</p>
</dd>
<dt>Returns:</dt><dd><p>df(Pandas dataframe): entire dataset with null patients dropped
and zeros as NaNs
df_meta(Pandas dataframe): dataframe of metabolite values (already
log transformed)
df_micro(Pandas dataframe): dataframe of microbiome counts</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.trimdata.calculate_shannon_diversity">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.trimdata.</span></span><span class="sig-name descname"><span class="pre">calculate_shannon_diversity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/trimdata.html#calculate_shannon_diversity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.trimdata.calculate_shannon_diversity" title="Permalink to this definition"></a></dt>
<dd><p>Calculates shannon (alpha) diversity index for each patient</p>
<blockquote>
<div><dl class="simple">
<dt>Parameters:</dt><dd><p>data(Pandas dataframe): microbiome df where each row is a patient
and .X is their microbiome count data</p>
</dd>
<dt>Returns:</dt><dd><p>shannon_diversity(Pandas df): 1-column df where each index is a
patient and the column is their shannon diversity value</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.trimdata.count_to_abundance">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.trimdata.</span></span><span class="sig-name descname"><span class="pre">count_to_abundance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/trimdata.html#count_to_abundance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.trimdata.count_to_abundance" title="Permalink to this definition"></a></dt>
<dd><p>Transform microbiome sequencing count data to relative abundance with
Center-log ratio (CLR) transformation.</p>
<blockquote>
<div><dl class="simple">
<dt>Parameters:</dt><dd><p>df(Pandas dataframe): microbiome counts per species with metadata
columns MGS count, Gene count, Microbial load, ID, and Status</p>
</dd>
<dt>Returns:</dt><dd><p>abund(Pandas dataframe): microbe center-log ratio abundances per
species with metadata columns ID and Status</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.trimdata.preprocess">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.trimdata.</span></span><span class="sig-name descname"><span class="pre">preprocess</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fill_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/trimdata.html#preprocess"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.trimdata.preprocess" title="Permalink to this definition"></a></dt>
<dd><p>Main preprocessing function that runs all others.</p>
<blockquote>
<div><p>Parameters: None</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>df_metamicro(Pandas dataframe): processed dataframe
micro_cols(list): feature columns of microbial species
metabo_cols(list): feature coumns of metabolites</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.trimdata.read_csvs">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.trimdata.</span></span><span class="sig-name descname"><span class="pre">read_csvs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/trimdata.html#read_csvs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.trimdata.read_csvs" title="Permalink to this definition"></a></dt>
<dd><p>Read raw data csv files from Cardio/data and merge into one large dataframe</p>
<blockquote>
<div><dl class="simple">
<dt>Parameters:</dt><dd><p>None</p>
</dd>
<dt>Returns:</dt><dd><p>all_joined(Pandas dataframe): dataframe containing metadata,
microbiome, and metabolome for each patient
metadata_columns(column index object): column names of metadata
microbiome_columns(column index object): column names of microbes
metabolome_columns(column index object): column names of
metabolites</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="cardio.preprocessing_scripts.trimdata.sparse_filt">
<span class="sig-prename descclassname"><span class="pre">cardio.preprocessing_scripts.trimdata.</span></span><span class="sig-name descname"><span class="pre">sparse_filt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">na_lim</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/cardio/preprocessing_scripts/trimdata.html#sparse_filt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#cardio.preprocessing_scripts.trimdata.sparse_filt" title="Permalink to this definition"></a></dt>
<dd><p>Filtering out sparse and uncharacterized features in a data subset</p>
<blockquote>
<div><dl class="simple">
<dt>Parameters:</dt><dd><p>df(Pandas dataframe): dataframe with feature columns and patients
as indices
remove_str(string): identifier of uncharacterized columns (e.g.
unknown metabolites or species) - either ‘X-’ or ‘unclassified’
here
na_lim(int): number of acceptable nas in feature column</p>
</dd>
<dt>Returns:</dt><dd><p>df(Pandas dataframe): trimmed dataframe</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

</section>
<section id="module-cardio.preprocessing_scripts">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-cardio.preprocessing_scripts" title="Permalink to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="cardio.html" class="btn btn-neutral float-left" title="cardio package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="cardio.tests.html" class="btn btn-neutral float-right" title="cardio.tests package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Kira Olander, Mattias Tolhurst, Paul Kim, Alyssa Easton, Maggie Cook.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>